starport-core/src/main/scala/com/krux/starport/StartScheduledPipelines.scala

```diff
@@ -218,35 +125,22 @@ object StartScheduledPipelines extends StarportActivity {
         new ForkJoinPool(parallel * Runtime.getRuntime.availableProcessors)
       )
 
+    val taskDispatcher: TaskDispatcher = conf.dispatcherType match {
+      case "default" => new TaskDispatcherImpl()
+      case x => throw new NotImplementedError(s"there is no task dispatcher implementation for $x")
+    }
+
+    // in case of default taskDispatcher: the dispatchedPipelines and succesfulPipelines metrics should be exactly same.
+    val dispatchedPipelines = metrics.register("counter.successful-pipeline-dispatch-count", new Counter())
     val successfulPipelines = metrics.register("counter.successful-pipeline-deployment-count", new Counter())
     val failedPipelines = metrics.register("counter.failed-pipeline-deployment-count", new Counter())
 
-    parPipelineModels.foreach { p =>
-
-      val timerInst = scheduleTimer.time()
-
-      logger.info(s"deploying pipleine ${p.name}")
-
-      val (status, output, pipelineName) = deployPipeline(
-        p, options.scheduledStart, options.scheduledEnd, localJars(p.jar))
-
-      Console.err.print(output)
-
-      if (status == 0) {  // deploy successfully, perform activation
-        activatePipeline(p, pipelineName, options.scheduledStart, options.actualStart, options.scheduledEnd)
-        logger.info(
-          s"deployed pipeline ${p.name} in ${TimeUnit.SECONDS.convert(timerInst.stop(), TimeUnit.NANOSECONDS)}"
-        )
-        successfulPipelines.inc()
-      } else {  // otherwise handle the failure and send notification
-        ErrorHandler.pipelineScheduleFailed(p, output)
-        logger.warn(
-          s"failed to deploy pipeline ${p.name} in ${TimeUnit.SECONDS.convert(timerInst.stop(), TimeUnit.NANOSECONDS)}"
-        )
-        failedPipelines.inc()
-      }
+    parPipelineModels.foreach(p => processPipeline(taskDispatcher, p, options, localJars(p.jar), dispatchedPipelines, failedPipelines))
 
-    }
+    // retrieve the scheduled pipelines and save the information in the db
+    val scheduledPipelines = taskDispatcher.retrieve(conf)
```

Comment:
in the case where dispatcher executed remotely, we need to make sure, we have received all dispatcher results before we start scheduling.

Reply:
> in the case where dispatcher executed remotely, we need to make sure, we have received all dispatcher results before we start scheduling.

Nice catch. Fixing.

Reply:
Not sure how the new change fixes the issue. When `retrieve` is called, it's not necessarily all pipelines have been retrieved by a potentially remote dispatcher, right?

Reply:
We are initializing TaskDispatcher (Line 116) before we fetch the pipelines from the database. 

A remote dispatcher will need to call retrievePipelines immediately after initialization. When we add a remote dispatcher, we will need to make sure that we do that.

 OR 

We can call retrieve pipelines for all dispatchers, it will be a no-op in case of local dispatcher.